### DISCUSSION TITLE
Multiple persistence units with WebSocket reactive messaging?

### DISCUSSION THREAD

**Discussion by Gonzalo-bia** [quarkusio/quarkus#28853] [Q&A] [answered]
Hello,

I have the following problem: I need to use reactive messaging (WebSockets) in combination with multiple datasources (Hibernate ORM). Due to having multiple datasources, I can't use HIbernate Reactive. But at the same time, when I try to use a Panache repository in an `@OnMessage` method (`@ServerEndpoint`-annotated class), I get the following error:

> io.quarkus.runtime.BlockingOperationNotAllowedException: You have attempted to perform a blocking operation on a IO thread. This is not allowed, as blocking the IO thread will cause major performance issues with your application. If you want to perform blocking EntityManager operations make sure you are doing it from a worker thread.

I've tried returning a Uni from my repository method and resolving it asyncrhonously, but I still have this error, as the Hibernate library is the regular one, not the reactive.

Is there a way to:
- Use multiple datasources Reactively? (I assume not)
- Use one datasource reactively and the other one as JDBC
- Manually create a Uni in a Panache repo and convince Hibernate ORM it's in an IO thread
- Make a WS endpoint work synchronously with Smallrye reactive messaging

`Quarkus version: 2.10.0.Final`

[a comment by an automated account is not shown]

**Comment by ozangunalp** [comment]
I think you are using the Java websocket API, not the reactive messaging, if you are using `@OnMessage` and `@ServerEndpoint`. 

If I am not mistaken Hibernate Reactive doesn't support multiple persistence units. This issue mentions it #21110.

You can use Hibernate ORM (not reactive) with multiple persistence units, but you can't call blocking API on an IO thread, as the exception tells.

If you add `@Blocking` annotation to your `@OnMessage` the method will be called on a worker thread instead of the IO thread.

**Reply by Gonzalo-bia** [reply]
Thank you for your answer.

> I think you are using the Java websocket API, not the reactive messaging, if you are using @OnMessage and @ServerEndpoint.

I am, yes. But this is not the source of the problem, right? I can't call blocking ops on a Smallrye thread.

> You can use Hibernate ORM (not reactive) with multiple persistence units, but you can't call blocking API on an IO thread, as the exception tells.

Is there a way to manually make calls in a Panache repository asynchronous? I've tried returning a Uni from my repository method and subscribing to it, but I still get the IO exception...

> If you add @Blocking annotation to your @OnMessage the method will be called on a worker thread instead of the IO thread.

I already tried that, but it complains about not having an `@Input` or `@Output` configured

**Reply by ozangunalp** [reply]
You should be using the `io.smallrye.common.annotation.Blocking`. the other is specific to reactive messaging.

**Reply by Gonzalo-bia** [reply]
I'm afraid the issue remains, I've tried annotating the methods involved in the entire flow, but I still get the `BlockingOperationNotAllowedException`...

**Comment by ozangunalp** [comment, marked as answer]
Yeah, turns out I was wrong, and `@Blocking` doesn't work in that case. I am sorry for the confusion.

You'd need to set  `quarkus.websocket.dispatch-to-worker=true` config in `application.properties` for websocket methods to be called on worker threads instead of IO threads.

You'd also need to add `@Transactional` to methods using Hibernate ORM.

**Reply by Gonzalo-bia** [reply]
it works! Brilliant, thanks
